### PR TITLE
core: Migrate InternalSubchannel to use SynchronizedContext 

### PR DESCRIPTION
**Pull request by creamsoup**
one of the main motivation was trying to fix a deadlock in OkHttpClientTransport.
FYI. InProcessTransport.start returns Runnable, so can't remove the runnable.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 1]
```diff
@@ -275,33 +246,19 @@ private void startNewTransport() {
...
-          syncContext.drain();
-        }
+        reconnectTask = null;
+        channelLogger.log(ChannelLogLevel.INFO, "CONNECTING after backoff");
+        gotoNonErrorState(CONNECTING);
+        startNewTransport();
       }
     }
 
+    syncContext.throwIfNotInThisSynchronizationContext();;
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Move it to the beginning of the method?

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 2]
```diff
@@ -313,20 +270,21 @@ public void run() {
         "TRANSIENT_FAILURE ({0}). Will reconnect after {1} ns",
         printShortStatus(status), delayNanos);
     Preconditions.checkState(reconnectTask == null, "previous reconnectTask is not done");
-    reconnectCanceled = false;
-    reconnectTask = scheduledExecutor.schedule(
+    reconnectTask = syncContext.schedule(
         new LogExceptionRunnable(new EndOfCurrentBackoff()),
         delayNanos,
-        TimeUnit.NANOSECONDS);
+        TimeUnit.NANOSECONDS,
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: We probably don't need to use `LogExceptionRunnable` anymore. It was necessary because the runnable was scheduled on `ScheduledExecutorService` which doesn't handle exceptions as we would like to (6a04022d7aceb73d9cf8096b107930eac0f6c477)

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 3]
```diff
@@ -335,43 +293,43 @@ void resetConnectBackoff() {
...
-      syncContext.drain();
-    }
+    });
   }
 
-  @GuardedBy("lock")
-  private void gotoNonErrorState(ConnectivityState newState) {
-    gotoState(ConnectivityStateInfo.forNonError(newState));
+  private void gotoNonErrorState(final ConnectivityState newState) {
+    syncContext.execute(new Runnable() {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: `gotoNonErrorState()` is only run from synContext. Why is this necessary?

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 4]
```diff
@@ -555,10 +492,10 @@ private static void checkListHasNoNulls(List<?> list, String msg) {
...
     public void transportReady() {
       channelLogger.log(ChannelLogLevel.INFO, "READY");
-      Status savedShutdownReason;
-      try {
-        synchronized (lock) {
-          savedShutdownReason = shutdownReason;
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          Status savedShutdownReason = shutdownReason;
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: We can safely use `shutdownReason` directly within this scope.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java (hunk 3 above)
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: iirc, this was hard to figure out. the SyncContext execution order is different than lock. because it doesn't allow the re-entrance.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java (hunk 1 above)
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: done.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 5]
```diff
@@ -156,10 +133,8 @@ protected void handleNotInUse() {
   @Nullable
```
[the remark this answers is not in the document]
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: done.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 6]
```diff
@@ -79,53 +77,33 @@
...
-  // 3. Every synchronized("lock") must be inside a try-finally which calls drain() in "finally".
   private final SynchronizationContext syncContext;
 
   /**
    * The index of the address corresponding to pendingTransport/activeTransport, or at beginning if
    * both are null.
    */
-  @GuardedBy("lock")
-  private Index addressIndex;
+  private volatile Index addressIndex;
```
[the remark this answers is not in the document]
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: good point. if we adapt more SynchronizedContext, it would be nice to have something like GuardedBy.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java (hunk 2 above)
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: right. syncContext is already handling exception.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java (hunk 4 above)
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: yes thanks.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java (hunk 3 above)
Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Did it cause any problem? Calling inline is closer to the previous behavior under lock.

Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: yes, 3 tests are failing (io.grpc.internal.ManagedChannelImplTest.{channelsAndSubchannels_oob_instrumented_fail,oobchannels,channelsAndSubchannels_oob_instrumented_success}.

Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: These tests fail because `transportReady()` delivers a `READY` state to the listener before updating  `activeTransport`. `OobChannel` calls `DelayedClientTransport.reprocess()` as soon as receiving a `READY`, which will fail because `activeTransport` is not yet set. Previous code got away with this issue because it called `syncContext.drain()` after `activeTransport` is set, but the root cause, which is transitioning to a new state prematurely, was already there. Using `syncContext.excute()` inside `gotoNonErrorState()` only covers up the root cause.

The correct implementation is to transition to a new state **after** all necessary state changes have been made. If you move all `gotoNonErrorState()` and `gotoState()` calls to after their prerequisite state changes, those tests will pass.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 7]
```diff
@@ -80,52 +78,36 @@
...
+   */
   private final SynchronizationContext syncContext;
 
   /**
    * The index of the address corresponding to pendingTransport/activeTransport, or at beginning if
    * both are null.
    */
-  @GuardedBy("lock")
-  private Index addressIndex;
+  private volatile Index addressIndex;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Why is this volatile? It is never changed. It seems there are reads from the index without synchronization: tostring, getAddressGroups (just missing throwIfNotInThisSynchronizationContext()?). That needs to be resolved.

Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: @zhangkun83 and I discussed this before, this was tricky one iirc. Those two methods are called from elsewhere (getAddressGroups is called from ManagedChannelImpl.SubchannelImpl and OobChannel). if it returns Future, we need to create a thread otherwise it will hang. Slightly delay for the getAddressGroups looked acceptable according to Kun. @zhangkun83, can you explain again?

Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: My fault. Making `addressIndex` volatile doesn't help. It should in fact be final.
`Index.addressGroups` can be made volatile to help `toString()` and `getAddressGroups()`. They can't have `throwIfNotInThisSynchronizationContext()` because they didn't have such requirement when they were first introduced.

Comment (ejona86) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: `Index.addressGroups` being volatile would fix it, although may be a bit confusing to see all the unnecessary volatile reads (making you think synchronization is going on). We could also keep another field that the same as `Index.addressGroups` except being volatile. In can be in InternalSubchannel (vs Index) if we'd like.

Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: @zhangkun83 that's definitely worth a comment.

Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: +1 for having a `volatile addressGroups` in `InternalSubchannel`.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 8]
```diff
@@ -86,8 +86,19 @@
   /**
    * The index of the address corresponding to pendingTransport/activeTransport, or at beginning if
    * both are null.
+   *
+   * <p>Note: any {@link Index#updateAddresses(List)} should also update {@link #addressGroups}.
    */
-  private volatile Index addressIndex;
+  private Index addressIndex;
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: This should be final.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java (hunk 7 above)
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: yeah just added the volatile accessor for `addressGroups`, PTAL.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 9]
```diff
@@ -86,8 +86,19 @@
...
+
+  /**
+   * A volatile accessor to {@link Index#getAddressGroups()}. There are few methods ({@link
+   * #getAddressGroups()} and {@link #toString()} access this value where they supposed to access
+   * in the {@link #syncContext}. Ideally {@link Index#getAddressGroups()} can be volatile, so we
+   * don't need to maintain this volatile accessor. Although, having this accessor can reduce
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: For posterity (it is fine to leave as-is): for me it was more of an argument for readability. With this approach the intent of the code is clearer.
